Fetch upstream

DawnStone opened the pull request.

petterinteon reviewed (commented):
I suggest changing line 358 to 
std::cout << "Randomizing with HL_RANDOM_WEIGHT_SEED, seed = " << seed_str << "\n";
so that we can pick up the "seed =" string in our surrounding scripts

abadams committed:
Reenable warning about unscheduled update definitions

and fix associated issues in the tests and apps.

This is an old warning that stopped triggering because it wasn't tested.
We should either remove it, fix the trigger conditions, or perhaps make
it an error. This PR fixes the trigger conditions and fixes all
instances of the warning in our tests and apps.

The warning triggers if you schedule some but not all of the update
definitions of a Func. It's to protect against the common error of only
scheduling the pure definition of something like a summation. The
warning can be suppressed by inserting a call to func.update(idx).

abadams committed:
Fix atomics test

abadams committed:
Add Stage::unscheduled()

rootjalex committed:
remove incorrect docs on widening_add (#6625)

alexreinking committed:
Make IRComparer consider nans to be less than non-nans. (#6626)

* Make IRComparer consider nans to be less than non-nans. Fixes #6624

* Fix test

* Respond to reviewer comments.

* Return -1 instead of +1

rootjalex committed:
Disallow `Type::narrow()` and `Type::widen()` from producing bitwidths between 1 and 8 bits (#6622)

* Disallow Type::narrow() and Type::widen() from producing bitwidths between 1 and 8 bits

* Narrowing a 1-bit type should error

vksnk committed:
Avoid double narrowing in widening_add/widening_sub if type is 8-bit (#6629)

rootjalex committed:
Wild match object is not foldable (#6623)

steven-johnson committed:
Python Bindings didn't allow for zero-D Funcs, ImageParams, Buffers (#6633)

* Python Bindings didn't allow for zero-D Funcs, ImageParams, Buffers

There were no overloads or tests for accessing the element of any of these in the zero-D case, and the obvious syntax (`[]`, to mirror C++ `()` in these cases) isn't legal in Python. To support this uncommon-but-necessary case, I'm proposing that we use the syntax `[None]`, which isn't pretty, but is less bad than other options I've considered so far. (Suggestions welcome.)

* Use [()] instead of [None]

steven-johnson committed:
Clean up python_binding Makefile (#6634)

* Clean up python_binding Makefile

Some of the tests with Generator dependencies had deps set up in a weird way that made some downstream work painful. Cleaned up. Also vastly reduced build-time noise.

* Update Makefile

* Add linker scripts for Python extensions

knzivid committed:
python_bindings: acquire GIL before printing (#6635)

rootjalex committed:
Clear bounds info on casts when value bounds are undefined (for overflow types) (#6640)

steven-johnson committed:
Convert most remaining Generators to prefer statically-dimensioned Inputs and Output where possible (#6641)

This is the same as #6620, except that it omits autoschedulers/adams2019/cost_model_generator.cpp (which is unusually complex and not yet settled as to whether the changes are welcome). Basically an attempt to land the uncontroversial parts.

steven-johnson committed:
Fix const-correctness in C/C++ backend (Issue #6636) (#6638)

in the Load handler, we need to emit the cast in the form `TYPE const *` rather than `const TYPE *`, as TYPE could be `void *`, and the const would bind in a way we don't want.

steven-johnson committed:
Fix UB in hannk FillWithRandom operation. (#6645)

A recent change to libc++ rejects std::uniform_int_distribution<uint8_t>
explicitly (https://reviews.llvm.org/D114920). This change prevents this file
to build with upcoming revisions of the C++ toolchain.

steven-johnson committed:
Allow profiler feature under wasm iff wasm_threads is enabled (#6643)

The profiler requires threads, but works fine when wasm_threads are enabled.

steven-johnson committed:
Update initialization of WABT `store` field to work with top-of-tree (#6649)

The copy and move assignment operators for Store are going away; initialize ours in the

LebedevRI committed:
[CMake] Deduplicate `Halide_LLVM_VERSION` and `LLVM_PACKAGE_VERSION` (#6646)

steven-johnson committed:
Fix apparent type in PR #6294 (#6653)

steven-johnson committed:
Eliminate some unnecessary clamping in ClampUnsafeAccesses (#6297) (#6654)

* Eliminate some unnecessary clamping in ClampUnsafeAccesses (#6297)

* Update ClampUnsafeAccesses.cpp

* Update ClampUnsafeAccesses.cpp

* Update ClampUnsafeAccesses.cpp

Commit:
Timer based profiler (#6642)

* Add support for timer interrupt based profiling, which is useful for bring up
on embedded ("bare metal") systems that may not have a full OS with threads.

* Update runtime_api file with new routines.

* Turn locking back on in timer based profiling case as it can be used
in multiprocessor situations. (Both as an option on systems where
threads would be fine and on embedded systems which don't have time
shared threads but cores are dedicated to the Halide threadpool.)

* Add target flag for timer profiling and extend performance_profiler
test to cover timer profiling.

steven-johnson committed:
Fix 'variable set but not used` warning/error (#6658)

Yes, some compilers complain about this.

steven-johnson committed:
Python Bindings: fix Python `bool` -> `Expr` implicit conversion (#6657)

* Python Bindings: fix Python `bool` -> `Expr` implicit conversion

It was implicitly converting to a Halide `Int(32)` literal of value 1, but we want it to match Halide's boolean type of `UInt(1)`

* Update basics.py

* wip

* tickle buildbots

steven-johnson committed:
Add optional runtime H::R::Buffer access checks (#6660)

* Add optional runtime H::R::Buffer access checks

This adds some optional `assert()` checks to HRB's `operator()` and friends. They are only enabled if `HALIDE_RUNTIME_BUFFER_CHECK_INDICES=1` is defined at compile time.

Also fixes errors found by enabling these assertions and running tests.

* Update fast_pow.cpp

* clang-format

* tickle buildbots

* tickle buildbots

steven-johnson committed:
Allow `make test_apps` to work with ASAN (#6659)

* Allow `make test_apps` to work with ASAN

With asan or tsan in the target, there is a space in the OPTIMIZE var, so it needs to be quoted.

* tickle buildbots

* tickle buildbots

* tickle buildbots

steven-johnson committed:
Add ldscript code for Python extensions in CMake (#6665)

* Add ldscript code for Python extensions in CMake

We added ldscripts to the Makefile for Python extensions (to restrict exported symbols to just the PyInit_foo symbol), but neglected to do so for CMake. This corrects that.

steven-johnson committed:
Remove the nobuild/partialbuildmethod tests from python_bindings/ (#6668)

They no longer serve a purpose and are redundant to other tests.

LebedevRI committed:
`-mtune=`/`-mcpu=` support for x86 AMD CPU's (#6655)

* `-mtune=`/`-mcpu=` support for x86 AMD CPU's

* Move processor tune into it's own enum, out of features

* clang-format

* Target: make Processor more optional

* Processor: add explanatory comments which CPU is what

* Drop outdated changes

* Make comments in Processor more readable / fix BtVer2 comment

* Target: don't require passing Processor

* Make processor more optional in the features string serialization/verification

* Address review notes

* Undo introduction of halide_target_processor_t

* Fix year for btver2/jaguar

shoaibkamil committed:
Fix GPU depredication/scalarization (#6669)

* Scalarize predicated Loads

* Cleanup

* Fix gpu_vectorize scalarization for D3D12

* Fix OpenCL scalarization

* Minor fixes

* Formatting

* Address review comments

* Move Shuffle impl to CodeGen_GPU_C class

* Extra space removal

Co-authored-by: Shoaib Kamil <kamil@adobe.com>

steven-johnson committed:
Allow PyPipeline and PyFunc to realize() scalar buffers (#6674)

LebedevRI committed:
Future-proof 'processor` to `tune processor` (#6673)

steven-johnson committed:
Fix ctors for Realization (#6675)

For vector-of-Buffers, the ctor took a non-const ref to the argument, which was weird and nonsensical. Replaced with a const-ref version and and an rvalue-ref version; it turns out that literally *all* of the internal calls were able to use the latter, trivially saving some copies.

LebedevRI committed:
`-mtune=native` CPU autodetection for AMD Zen 3 CPU (#6648)

* `-mtune=native` CPU autodetection for AMD Zen 3 CPU

* Address review notes.

* Fix MSVC build

* Address review notes

alexreinking committed:
Bump development Halide version to 15.0.0 (#6678)

* Bump development Halide version to 15.0.0

* trigger buildbots

steven-johnson committed:
Clean up Python extensions in python_bindings (#6670)

* Remove the nobuild/partialbuildmethod tests from python_bindings/

They no longer serve a purpose and are redundant to other tests.

* WIP

* Update pystub.py

* wip

* wip

* wip

* Update TargetExportScript.cmake

* Update PythonExtensionHelpers.cmake

* PyExtensionGen didn't handle zero-dimensional buffers

steven-johnson committed:
Halide::Tools::save_image() should accept buffers with `const` types (#6679)

steven-johnson committed:
Fix "set but not used" warnings/errors (#6683)

* Fix "set but not used" warnings/errors

Apparently XCode 13.3 has smarter warnings about unused code and emits warnings/errors for these, so let's clean them up.

* Also fix missing `ssize_t` usage

steven-johnson committed:
Remove deprecated `Halide::Output` type (#6685)

It was deprecated (in favor of `OutputFileType` in Halide 14; let's remove it entirely for Halide 15.

steven-johnson committed:
Remove deprecated `build()` support from Generators (#6684)

This was deprecated in Halide 14; let's remove it entirely for Halide 15.

steven-johnson committed:
Drop support for LLVM12 (#6686)

* Drop support for LLVM12

Halide 15 only needs to support LLVM13 and LLVM13. Drop all the special-casing for LLVM12.

* Update packaging.yml

* Update presubmit.yml

* 13

* more

* Update presubmit.yml

* woo

* Update presubmit.yml

* Update run-clang-tidy.sh

* Update run-clang-tidy.sh

* Update .clang-tidy

* Update .clang-tidy

* wer

* Update Random.cpp

* wer

* sdf

* sdf

* Update packaging.yml

steven-johnson committed:
Upgrade to clang-format 13 (#6689)

Goal here: eliminate the need for a local version of llvm/clang-12, and don't stay too far behind the toolchain.

As always, clang-format doesn't promise backwards compatibility, but the main differences in formatting are:
- more regularization of spaces at the start of comments (I like this change)
- minor difference of formatting of function-pointer-type declarations (not a fan of this, but I can't find a way to disable it and it's only really used in a handful of place in the Python bindings)

steven-johnson committed:
Always mark _ucon as 'unused' in Codegen_C (#6691)

* Always mark _ucon as 'unused' in Codegen_C, even if asserts are enabled, since generated closure functions may not use it

* halide_unused -> halide_maybe_unused

* fix test_internal

* More halide_unused -> halide_maybe_unused

steven-johnson committed:
Add `break` to avoid 'possible unintentional fallthru' warning (#6694)

steven-johnson committed:
Silence "unknown warning" in Clang 13 (#6693)

Clang 13 removed the `return-std-move-in-c++11` warning entirely, so specifying it now warns that the warning is unknown.

steven-johnson committed:
Fixes for top-of-tree LLVM (#6697)

rootjalex committed:
Faster `widening_mul(int16x, int16x) -> int32x` for x86 (AVX2 and SSE2) (#6677)

* add widening_mul using vpmaddwd for AVX2

* add vpmaddwd/pmaddwd test

* add widening_mul with pmaddwd for SSE2

steven-johnson committed:
Remove deprecated versions of Func::prefetch() (#6698)

steven-johnson committed:
Remove deprecated JIT handler setters (#6699)

steven-johnson committed:
Drop support for Matlab extensions (#6696)

* Drop support for Matlab extensions

Anecdotally, this hasn't been used in ~years, and the original author (@dsharletg) had suggested dropping it a while back. I'm going to propose we go ahead and drop it for Halide 15 and see who complains.

* Fixes for top-of-tree LLVM

* Update force_include_types.cpp

* trigger buildbots

* Update CodeGen_LLVM.cpp

abadams committed:
llvm no longer wants a type suffix on vst intrinsics (#6701)

* llvm no longer wants a type suffix on vst intrinsics

* Fix silly mistake

* Change 64-bit only

Co-authored-by: Andrew Adams <anadams@adobe.com>

steven-johnson committed:
Python: make Func implicitly convertible to Stage (#6702) (#6704)

This allows for `compute_with` and `rfactor` to work more seamlessly in Python.

Also:
- Move two compute_with() variant bindings from PyFunc and PyStage to PyScheduleMethods, as they are identical between the two
- drive-by removal of redundant `py::implicitly_convertible<ImageParam, Func>();` call

steven-johnson committed:
Fix type-mangling for vst on arm32 for LLVM15 (#6705)

steven-johnson committed:
Remove the last remaining call to getPointerElementType() (#6715)

* Remove the last remaining call to getPointerElementType()

LLVM is moving to opaque pointers, we must have missed this one in previous work

* ARM vst mangling needs to be conditional on opaque ptrs

The fixes from last week regarding mangling of arm vst intrinsics needs to be made conditional on whether the pointer is opaque or not; this will change based on whether `-D CLANG_ENABLE_OPAQUE_POINTERS=ON|OFF` is defined when LLVM is built, but should be sniffed via this API, according to my LLVM contact.

* Revert "ARM vst mangling needs to be conditional on opaque ptrs"

This reverts commit 9901314ff75dd0bf651b23d09c1d1f5f07d49ffd.

steven-johnson committed:
ARM vst mangling needs to be conditional on opaque ptrs (#6716)

The fixes from last week regarding mangling of arm vst intrinsics needs to be made conditional on whether the pointer is opaque or not; this will change based on whether `-D CLANG_ENABLE_OPAQUE_POINTERS=ON|OFF` is defined when LLVM is built, but should be sniffed via this API, according to my LLVM contact.

steven-johnson committed:
Combine string constants in combine_strings() (#6717)

* Combine string constants in combine_strings()

This is a pretty trivial optimization, but when printing (or enabling `debug`), it cuts the number of `halide_string_to_string()` calls we generate by ~half.

* Update IROperator.cpp

steven-johnson committed:
Update CodeGen_PTX_Dev to use new PassManager (#6718)

* Update CodeGen_PTX_Dev to use new PassManager

This was still using the LegacyPassManager for optimization, which will be going away at some point. (Code changes by @alinas; I'm just opening this PR on her behalf)

* Fixes after review

steven-johnson committed:
Closure functions for parallel tasks should be internal, not external (#6720)

Minor optimization.

steven-johnson committed:
Smarten type_of<> for fn ptrs; fix async_parallel for C backend (#6719)

* Smarten type_of<> for fn ptrs; fix async_parallel for C backend

(Fixes #2093)

This basically just adds the right type annotations to make the parallel code produced by the C backend compile properly. This could have been fixed by inserted some brute-force void* casting into the C backend, but this felt a lot cleaner.

The one thing here I'm a little unsure about is how I extended the Type code to be able to handle function-pointer types correctly; it works but doesn't feel very elegant.

* Update Makefile

* Update LowerParallelTasks.cpp

* FunctionTypedef

steven-johnson committed:
Remove legacy::FunctionPassManager usage in Codegen_PTX_Dev (#6722)

LLVM devs indicate that none of the passes in this usage actually do anything and it can be safely removed.

LebedevRI committed:
`get_amd_processor()`: implement detection for the rest of supported AMD CPU's (#6711)

I have *not* personally tested that these are detected correctly,

Cross-reference between
* https://github.com/llvm/llvm-project/blob/955cff803e081640e149fed0742f57ae1b84db7d/llvm/lib/Support/Host.cpp#L968-L1041
* https://github.com/llvm/llvm-project/blob/955cff803e081640e149fed0742f57ae1b84db7d/compiler-rt/lib/builtins/cpu_model.c#L520-L586
* https://github.com/gcc-mirror/gcc/blob/000c1b89d259fadb466e1f2e63c79da45fd17372/gcc/common/config/i386/cpuinfo.h#L111-L264

steven-johnson committed:
Add Func::output_type() method (#6724)

* Add Func::output_type() method

* Add Python

steven-johnson committed:
Grab-bag of minor Python fixes (#6725)

steven-johnson committed:
Deprecate variadic-template version of Realization ctor (#6695)

* Deprecate variadic-template version of Realization ctor

The variadic-template approach was useful before C++11 (!) added brace initialization, but preferring an explicit vector-of-Buffer is arguably better, and provides better symmetry with the Python bindings.

Also, some drive-by tweaks to other Realization methods.

* Update PyPipeline.cpp

* trigger buildbots

rootjalex committed:
Remove `rounding_halving_sub` and non-existent arm rhsub instructions (#6723)

* remove arm (s | u)rhsub instructions

* remove rounding_halving_sub intrinsic entirely

steven-johnson committed:
Augment Halide::Func to allow for constraining Type and Dimensionality (#6734)

This enhances Func by allowing you to (optionally) constrain the type(s) of Exprs that the Func can contain, and/or the dimensionality of the Func. (Attempting to violate either of these will assert-fail.)

There are a few goals here:
- Enhanced code readability; in cases where a Func's values may not be obvious from the code flow, this can allow an in-code way of declaring it (rather than via comments)
- Enhanced type enforcement; specifying constraints allows us to fail in type-mismatched compilations somewhat sooner, with somewhat better error messages.
- Better symmetry for AOT/JIT code generation with ImageParam, in which the inputs (ImageParam) have a way to specify the required concrete type, but the outputs (Funcs) don't.

If this is accepted, then subsequent changes will likely add uses where it makes sense (e.g., the Func associated with an ImageParam should always have both type and dimensionality specified since it will always be well-known).

Note that this doesn't add any C++ template class for static declarations (e.g. `FuncT<float, 2>` -> `Func(Float(32), 2)`); these could be added later if desired.

steven-johnson committed:
More typed-Func work (#6735)

- Allow Func output_type(s)(), outputs(), dimensions(), and output_buffer(s)() to be called on undefined Funcs if the Func has required_type and required_dimensions specified. This allows for greater flexibility in defining pipelines in which you may want to set or examine constraints on a Func that hasn't been defined yet; previously this required restructuring code or other awkwardness.
- Ensure that the Funcs that are defined for ImageParams and Generator fields define the types-and-dims when known.
- Add some tests.

steven-johnson committed:
Add missing #include <functional> in ThreadPool.h (#6738)

* Add missing #include <function> in ThreadPool.h

* Update ThreadPool.h

steven-johnson committed:
Fix regression from #6734 (#6739)

That change inadvertently required the RHS of an update stage that used `+=` (or similar operators) to match the LHS type, which should be required (implicit casting of the RHS is expected). Restructured to remove this, but still ensure that auto-injection of a pure definition matches the required types (if any), and updated tests.

steven-johnson committed:
Add forwarding for the recently-added Func::output_type() method (#6741)

steven-johnson committed:
Silence "unscheduled update stage" warnings in msan_generator.cpp (#6740)

steven-johnson committed:
Add __pycache__ to toplevel .gitignore file (#6743)

steven-johnson committed:
Revise PyStub calling convention for GeneratorParams (#6742)

This is a rethink of https://github.com/halide/Halide/pull/6661, trying to make it saner in anticipation of the ongoing Python Generator work.

TL;DR: instead of mixing GeneratorParams in with the rest of the keywords, segregate them into an optional `generator_params` keyword argument, which is a plain Python dict. This neatly solves a couple of problems:
- synthetic params with funky names aren't a problem anymore.
- error reporting is simpler because before an unknown keyword could have been intended to be a GP or an Input.
- GP values are now clear and distinct from Inputs, which is IMHO a good thing.

This is technically a breaking change, but I doubt anyone will notice; this is mainly here to get a sane convention in place for use with Python Generators as well.

Also, a drive-by change to Func::output_types() to fix the assertion error message.
